File: scripts/seed-scav.py
# -*- coding: utf-8 -*-
"""Seed scav_* tables from data/seed.json."""
from __future__ import annotations

import json
from pathlib import Path
from urllib.parse import quote_plus

ROOT = Path(__file__).resolve().parents[1]
SEED = ROOT / "data" / "seed.json"
ENV = Path(r"C:\Users\cyber\Projects\fleetmind\.env.local")


def main() -> None:
    import psycopg

    vals = {}
    for line in ENV.read_text(encoding="utf-8").splitlines():
        if "=" in line and not line.strip().startswith("#"):
            k, v = line.split("=", 1)
            vals[k.strip()] = v.strip().strip('"').strip("'")
    ref = vals["SUPABASE_PROJECT_REF"]
    password = vals["SUPABASE_DB_PASSWORD"]
    data = json.loads(SEED.read_text(encoding="utf-8"))
    conninfo = (
        f"postgresql://postgres.{ref}:{quote_plus(password)}"
        f"@aws-0-ca-central-1.pooler.supabase.com:6543/postgres"
    )
    n_players = 0
    n_comps = 0
    with psycopg.connect(conninfo) as conn:
        for p in data.get("players") or []:
            name = p["display_name"]
            row = conn.execute(
                "select id from public.scav_players where lower(display_name)=lower(%s)",
                (name,),
            ).fetchone()
            if not row:
                row = conn.execute(
                    "insert into public.scav_players (display_name) values (%s) returning id",
                    (name,),
                ).fetchone()
                n_players += 1
            player_id = row[0]
            for c in p.get("completions") or []:
                cur = conn.execute(
                    """
                    insert into public.scav_completions
                      (player_id, period, period_label, map, source)
                    values (%s,%s,%s,%s,'scan_seed')
                    on conflict (player_id, period, map) do nothing
                    returning id
                    """,
                    (player_id, c["period"], c["period_label"], c["map"]),
                )
                if cur.fetchone():
                    n_comps += 1
        conn.commit()
    print(f"New players={n_players} new completions={n_comps}")


if __name__ == "__main__":
    main()

File: scripts/seed-scav.js
/**
 * Seed scav_* from data/seed.json via FleetMind DB password.
 */
const fs = require("fs");
const path = require("path");
const { Client } = require("pg");

const envPath = path.join("C:/Users/cyber/Projects/fleetmind/.env.local");
const seedPath = path.join(__dirname, "..", "data", "seed.json");

const env = Object.fromEntries(
  fs
    .readFileSync(envPath, "utf8")
    .split(/\r?\n/)
    .filter((l) => l.includes("=") && !l.trim().startsWith("#"))
    .map((l) => {
      const i = l.indexOf("=");
      return [l.slice(0, i).trim(), l.slice(i + 1).trim()];
    })
);

const pass = encodeURIComponent(env.SUPABASE_DB_PASSWORD);
const ref = env.SUPABASE_PROJECT_REF;
const connectionString = `postgresql://postgres:${pass}@db.${ref}.supabase.co:5432/postgres`;
const data = JSON.parse(fs.readFileSync(seedPath, "utf8"));

(async () => {
  const c = new Client({ connectionString, ssl: { rejectUnauthorized: false } });
  await c.connect();
  let nPlayers = 0;
  let nComps = 0;
  for (const p of data.players || []) {
    const name = p.display_name;
    let row = await c.query(
      "select id from public.scav_players where lower(display_name)=lower($1)",
      [name]
    );
    let playerId;
    if (!row.rows.length) {
      row = await c.query(
        "insert into public.scav_players (display_name) values ($1) returning id",
        [name]
      );
      nPlayers += 1;
      playerId = row.rows[0].id;
    } else {
      playerId = row.rows[0].id;
    }
    for (const comp of p.completions || []) {
      const ins = await c.query(
        `insert into public.scav_completions
           (player_id, period, period_label, map, source)
         values ($1,$2,$3,$4,'scan_seed')
         on conflict (player_id, period, map) do nothing
         returning id`,
        [playerId, comp.period, comp.period_label, comp.map]
      );
      if (ins.rows.length) nComps += 1;
    }
  }
  const board = await c.query(
    `select count(*)::int as players from public.scav_players`
  );
  const comps = await c.query(
    `select count(*)::int as completions from public.scav_completions`
  );
  console.log(
    `seeded new_players=${nPlayers} new_completions=${nComps} totals players=${board.rows[0].players} completions=${comps.rows[0].completions}`
  );
  await c.end();
})().catch((e) => {
  console.error(e.message || e);
  process.exit(1);
});
